### PR TITLE
Feature/add more command managers

### PR DESCRIPTION
**Pull request by vybhav-ibr**
Adds the position command manager and the pose command manager and the accompanying reward functions for tracking

### CODE REVIEW

File: genesis_forge/managers/command/pose_command.py [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+    Lifecycle Operations
+    """
+
+    def resample_command(self, env_ids: list[int]):
+        """
+        Overwrites commands for environments that should be standing still.
+        """
+        super().resample_command(env_ids)
+        if not self.enabled:
+            return
```
Comment (jgillick) [inline genesis_forge/managers/command/pose_command.py]: What's the purpose of this function if it's just calling the parent function?

File: genesis_forge/managers/command/position_command.py [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+    Lifecycle Operations
+    """
+
+    def resample_command(self, env_ids: list[int]):
+        """
+        Overwrites commands for environments that should be standing still.
+        """
+        super().resample_command(env_ids)
+        if not self.enabled:
+            return
```
Comment (jgillick) [inline genesis_forge/managers/command/position_command.py]: This seems unnecessary

File: genesis_forge/managers/command/position_command.py [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+    "sphere_offset": 0.03,
+    "sphere_radius": 0.02,
+    "commanded_color": (0.0, 0.5, 0.0, 1.0),
+}
+
+
+class PositionCommandManager(CommandManager):
+    """
+    Generates a position command from uniform distribution.
+    The command comprises of a linear velocity in x and y direction and an angular velocity around the z-axis.
```
Comment (jgillick) [inline genesis_forge/managers/command/position_command.py]: Is this comment correct?

File: genesis_forge/managers/command/position_command.py [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+from .command_manager import CommandManager, CommandRangeValue
+
+
+THIS_DIR = os.path.dirname(os.path.abspath(__file__))
+
+
+class PositionCommandRange(TypedDict):
+    pos_x: CommandRangeValue
+    pos_y: CommandRangeValue
+    pos_z: CommandRangeValue
```
Comment (jgillick) [inline genesis_forge/managers/command/position_command.py]: You can simplify these to: x, y, z

File: genesis_forge/managers/command/position_command.py [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+        class MyEnv(GenesisEnv):
+            def config(self):
+                # Create a velocity command manager
+                self.position_command_manager = PositionCommandManager(
+                    self,
+                    visualize=True,
+                    range = {
+                        "pos_x_range": (-5.0, 5.0),
+                        "pos_y_range": (-5.0, 5.0),
+                        "pos_z_range": (0.29, 0.31),
```
Comment (jgillick) [inline genesis_forge/managers/command/position_command.py]: This doesn't seem to match what's defined in `PositionCommandRange`

File: genesis_forge/managers/command/position_command.py [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+            return
+        super().step()
+        self._render_sphere()
+
+    def use_gamepad(
+        self,
+        gamepad: Gamepad,
+        pos_x_axis: int = 0,
+        pos_y_axis: int = 1,
+        pos_z_axis: int = 2,
```
Comment (jgillick) [inline genesis_forge/managers/command/position_command.py]: x, y, z

File: genesis_forge/managers/command/position_command.py [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+            pos_z_axis: Map this gamepad axis index to the position in the z-direction.
+        """
+        super().use_gamepad(
+            gamepad,
+            range_axis={
+                "pos_x": pos_x_axis,
+                "pos_y": pos_y_axis,
+                "pos_z": pos_z_axis,
+            },
+        )
```
Comment (jgillick) [inline genesis_forge/managers/command/position_command.py:173]: Have you tested this?

File: genesis_forge/managers/command/position_command.py [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+    def _render_sphere(self):
+        """
+        Render the command sphere showing position commands.
+
+        The commanded position sphere (green) shows the position in the world frame
+        """
+        if not self.debug_visualizer:
+            return
+
+        # Remove existing arrows
```
Comment (jgillick) [inline genesis_forge/managers/command/position_command.py]: arrows?

File: genesis_forge/managers/command/position_command.py [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+        for sphere in self._sphere_nodes:
+            self.env.scene.clear_debug_object(sphere)
+        self._sphere_nodes = []
+
+        for i in self.debug_envs_idx:
+            # Target arrow (robot-relative command transformed to world coordinates for visualization)
```
Comment (jgillick) [inline genesis_forge/managers/command/position_command.py]: Update the comment

File: genesis_forge/managers/command/position_command.py [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+                color=self.visualizer_cfg["commanded_color"],
+            )
+
+    def _draw_sphere(
+        self,
+        pos: torch.Tensor,
+        color: list[float],
+    ):
+        try:
+            node = self.env.scene.draw_debug_sphere(
```
Comment (jgillick) [inline genesis_forge/managers/command/position_command.py]: Using `draw_debug_spheres` to draw all spheres at the same time will be much more performant. Unfortunately, we cannot do that for arrows.

File: genesis_forge/mdp/rewards.py [hunk 11]
```diff
@@ -282,6 +285,142 @@ def action_rate_l2(env: GenesisEnv) -> torch.Tensor:
     return torch.sum(torch.square(last_actions - actions), dim=1)
 
 
+"""
+Position Command Rewards
+"""
+
+
+def commonad_tracking_position(
```
Comment (jgillick) [inline genesis_forge/mdp/rewards.py]: Typo: `commonad` -> `command`

File: genesis_forge/mdp/rewards.py [hunk 12]
```diff
@@ -282,6 +285,142 @@ def action_rate_l2(env: GenesisEnv) -> torch.Tensor:
...
+        link = entity.links[0]
+    else:
+        try:
+            link = entity.get_link(link_name)
+        except Exception as e:
+            raise ValueError(
+                f"Link name '{link_name}' not found in entity '{entity_attr}'."
+            )
+
+    pos = link.get_pos()
```
Comment (jgillick) [inline genesis_forge/mdp/rewards.py:335]: Why not just use `entity.get_pos()`?

File: genesis_forge/mdp/rewards.py [hunk 13]
```diff
@@ -282,6 +285,142 @@ def action_rate_l2(env: GenesisEnv) -> torch.Tensor:
...
+    link_name: str = None,
+    entity_manager: "EntityManager" = None,
+) -> torch.Tensor:
+    """
+    Penalize base pose away from target (both position and orientation) with separate sensitivities.
+
+    Args:
+        env: The Genesis environment containing the robot
+        command: The commanded XYZ position and Euler angles (in world frame), shape (num_envs, 6) where first 3 are position and last 3 are Euler angles.
+        position_cmd_manager: The velocity command manager
```
Comment (jgillick) [inline genesis_forge/mdp/rewards.py]: pos_sensitivity

File: genesis_forge/mdp/rewards.py [hunk 14]
```diff
@@ -282,6 +285,142 @@ def action_rate_l2(env: GenesisEnv) -> torch.Tensor:
...
+    link_name: str = None,
+    entity_manager: EntityManager = None,
+) -> torch.Tensor:
+    """
+    Penalize base pos away from target.
+
+    Args:
+        env: The Genesis environment containing the robot
+        command: The commanded XYZ position the the world frame, its shape is(num_envs, 3)
+        position_cmd_manager: The velocity command manager
```
Comment (jgillick) [inline genesis_forge/mdp/rewards.py]: "The velocity command manager"?

File: genesis_forge/mdp/rewards.py [hunk 15]
```diff
@@ -282,6 +285,142 @@ def action_rate_l2(env: GenesisEnv) -> torch.Tensor:
...
+
+    link_pos = link.get_pos()
+    link_euler = quat_to_xyz(link.get_quat())
+
+    if pose_cmd_manager is not None:
+        command_pos = pose_cmd_manager.command[:, :3]
+        command_euler = pose_cmd_manager.command[:, 3:6]
+    else:
+        command_pos = command[:, :3]
+        command_euler = command[:, 3:6]
```
Comment (jgillick) [inline genesis_forge/mdp/rewards.py]: Duplicate statements. Instead, assign to the same variable and then pull the values out.

Something like this:
```python
if pose_cmd_manager is not None:
        command = pose_cmd_manager.command
command_pos = command[:, :3]
command_euler = command[:, 3:6]
```

File: genesis_forge/mdp/rewards.py [hunk 16]
```diff
@@ -282,6 +285,142 @@ def action_rate_l2(env: GenesisEnv) -> torch.Tensor:
...
+        command is not None or pose_cmd_manager is not None
+    ), "Either command or position_cmd_manager must be provided to command_tracking_base_pose"
+
+    if entity_manager is not None:
+        entity = entity_manager.entity
+    else:
+        entity = getattr(env, entity_attr)
+
+    if link_name is None or link_name == "":
+        link = entity.links[0]
```
Comment (jgillick) [inline genesis_forge/mdp/rewards.py]: `link = entity.base_link`

File: genesis_forge/managers/command/position_command.py [hunk 17]
```diff
@@ -0,0 +1,229 @@
...
+import torch
+import genesis as gs
+
+from genesis_forge.genesis_env import GenesisEnv
+from genesis_forge.gamepads import Gamepad
+
+from .command_manager import CommandManager, CommandRangeValue
+
+
+THIS_DIR = os.path.dirname(os.path.abspath(__file__))
```
Comment (jgillick) [inline genesis_forge/managers/command/position_command.py]: Why is this necessary?

File: genesis_forge/managers/command/pose_command.py [hunk 18]
```diff
@@ -0,0 +1,247 @@
...
+import genesis as gs
+from genesis.utils.geom import euler_to_R
+import numpy as np
+from genesis_forge.genesis_env import GenesisEnv
+from genesis_forge.gamepads import Gamepad
+
+from .command_manager import CommandManager, CommandRangeValue
+
+
+THIS_DIR = os.path.dirname(os.path.abspath(__file__))
```
Comment (jgillick) [inline genesis_forge/managers/command/pose_command.py]: Why is this necessary?

File: genesis_forge/managers/command/pose_command.py [hunk 19]
```diff
@@ -0,0 +1,247 @@
...
+    Generates a pose command from uniform distribution.
+    The command comprises of a (x,y,z) position and (x,y,z) euler angles
+
+    IMPORTANT: The position commands are interpreted as world-relative coordinates:
+    - pos-X-axis: x coordinate of the target position
+    - pos-Y-axis: y coordinate of the target position
+    - pos-Z-axis: z coordinate of the target position
+    - euler-X-axis: x coordinate of the target orientation
+    - euler-Y-axis: y coordinate of the target orientation
+    - euler-Z-axis: z coordinate of the target orientation
```
Comment (jgillick) [inline genesis_forge/managers/command/pose_command.py]: These are not accurate to `PoseCommandRange`

File: genesis_forge/managers/command/pose_command.py [hunk 20]
```diff
@@ -0,0 +1,247 @@
...
+        class MyEnv(GenesisEnv):
+            def config(self):
+                # Create a velocity command manager
+                self.pose_command_manager = PoseCommandManager(
+                    self,
+                    visualize=True,
+                    range = {
+                        "pos_x_range": (-5.0, 5.0),
+                        "pos_y_range": (-5.0, 5.0),
+                        "euler_z_range": (-1.57, 1.57),
```
Comment (jgillick) [inline genesis_forge/managers/command/pose_command.py]: This is not accurate to `PoseCommandRange`

File: genesis_forge/managers/command/pose_command.py [hunk 21]
```diff
@@ -0,0 +1,247 @@
...
+    ):
+        super().__init__(env, range=range, resample_time_sec=resample_time_sec)
+        self._sphere_nodes: list = []
+        self.debug_visualizer = debug_visualizer
+        self.visualizer_cfg = {**DEFAULT_VISUALIZER_CONFIG, **debug_visualizer_cfg}
+        self.debug_envs_idx = None
+
+        self._is_standing_env = torch.zeros(
+            env.num_envs, dtype=torch.bool, device=gs.device
+        )
```
Comment (jgillick) [inline genesis_forge/managers/command/pose_command.py]: This seems unused.

File: genesis_forge/managers/command/position_command.py [hunk 22]
```diff
@@ -0,0 +1,229 @@
...
+    ):
+        super().__init__(env, range=range, resample_time_sec=resample_time_sec)
+        self._sphere_nodes: list = []
+        self.debug_visualizer = debug_visualizer
+        self.visualizer_cfg = {**DEFAULT_VISUALIZER_CONFIG, **debug_visualizer_cfg}
+        self.debug_envs_idx = None
+
+        self._is_standing_env = torch.zeros(
+            env.num_envs, dtype=torch.bool, device=gs.device
+        )
```
Comment (jgillick) [inline genesis_forge/managers/command/position_command.py]: This seems unused

File: genesis_forge/managers/command/pose_command.py [hunk 23]
```diff
@@ -0,0 +1,247 @@
...
+        """
+        super().use_gamepad(
+            gamepad,
+            range_axis={
+                "pos_x": pos_x_axis,
+                "pos_y": pos_y_axis,
+                "pos_z": pos_z_axis,
+                "euler_x": euler_x_axis,
+                "euler_y": euler_y_axis,
+                "euler_z": euler_z_axis,
```
Comment (jgillick) [inline genesis_forge/managers/command/pose_command.py]: I don't think this will work as most gamepads only have 2 joysticks, but this is expecting 3. Have you tested this?

**Review by jgillick — CHANGES_REQUESTED** [review]

**Comment by jgillick** [comment]
These require functional examples that converge.

File: genesis_forge/managers/command/pose_command.py (hunk 23 above)
Comment (vybhav-ibr) [reply genesis_forge/managers/command/pose_command.py]: i have not tested this, is it possible to set pos, click a button and then set orientation ?. if it is possible, please make the necessary changes. if this is not possible(or it doesn't feel intuitive), what would you suggest as the alternative ?

File: genesis_forge/mdp/rewards.py (hunk 12 above)
Comment (vybhav-ibr) [reply genesis_forge/mdp/rewards.py:335]: i want the user to be able to use this command manager with any link, that way this reward function can be used for manipulation examples as well. Also, sometimes the base_link may not be the actual center of the robot

**Comment by vybhav-ibr** [comment]
the example i included does converge, but not very well, further tuning is required. could you help with the reward engineering ?. The main issue I am facing is that, when the command position is more than half a meter from the current position the robot just gives up and falls on the ground, I have tried increasing the timeout, increase the reward weight, increase sensitivity. But I couldn't get a satisfactory result, 

https://github.com/user-attachments/assets/f3df1780-89f6-42e8-be6f-4921eb664e8c
The jitter you see in between the two successful targets is the repeated resets,

**Comment by jgillick** [comment]
> the example i included does converge, but not very well

Are these new managers for actual environments/robots you're working with that _do_ converge? If so, the example doesn't have to use the Go2 robot, and you can import a robot that you're working with as the example. 

Both examples are for the position command, and neither seems to demonstrate the pose command. Have you verified that PoseCommandManager works?

Also, you copied the README files without updating their contents. Please stop blindly copying code without updating it for your usecase.

File: genesis_forge/managers/command/pose_command.py (hunk 23 above)
Comment (jgillick) [reply genesis_forge/managers/command/pose_command.py]: Why are you submitting code that you have not tested?
